### PR TITLE
[CHW-35] Tech Spec

### PR DESCRIPTION
**Pull request by denisshevchenko**
This is a skeleton of Tech Spec, [CHW-35](https://iohk.myjetbrains.com/youtrack/issue/CHW-35).

### CODE REVIEW

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+fully concurrently? I mean publishing to the network, storing in the `wallet-db`, etc.
+
+## Scalability
+
+Initially clusters of nodes will be hosted by IOHK (on AWS EC2). And we need a mechanism of auto-scaling of the
+cluster: it must be able to grow proportionally to growing number of clients.
+
+**TODO**: To think about auto-scale supervisor. Suppose we have 1 cluster that consists of 100 nodes and each of node
+can reliably serve 100 clients. If supervisor sees that number of clients connected to this cluster is growing (and
+becomes more than 10000), it launches additional nodes.
```
Comment (denisshevchenko) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: @domenkozar What do you think about it?

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+Currently it is impossible, because transaction contains inputs, outputs and attributes, and we have to
+know _previous_ transaction to calculate a fee. So we must change transaction's format for explicit
+including a fee in it.
+
+**TODO**: To think about it: should the fee be a part of transaction's attributes, or it should be a
+separate field?
+
+### Other Attacks
+
+**TODO**: To think about it.
```
Comment (denisshevchenko) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: @vincenthz What do you think about other possible attacks?

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+After successful TLS checking client and server establish a new **session**.
+
+**TODO**: To think what is a session? What data is associated with the session? Does it have any timeout?
+
+## Crypto Operations
+
+A description of how the main operations on a crypto-wallet will be implemented, according to the final
+index derivation mechanism chosen (sequential vs random).
+
+**TODO**: To think about it.
```
Comment (denisshevchenko) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: @vincenthz I know almost nothing about it, so I need your advice.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+### HTTPS
+
+Client must be sure that server is not a malicious one, so we use HTTPS-connection with a standard check
+of TLS certs on the client side.
+
+### Session
+
+After successful TLS checking client and server establish a new **session**.
+
+**TODO**: To think what is a session? What data is associated with the session? Does it have any timeout?
```
Comment (denisshevchenko) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: @adinapoli-iohk @dcoutts I need an advice about session.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+this cluster consists of 100 nodes, each node has to be able to work with 100 clients simultaneously and reliably.
+Thus, the longest procedure is a new payment (creating new transaction), and even if all these 100 clients will
+sign their transactions and send it for publishing at the same time, the node must be able to handle these
+transactions reliably and as fast as possible.
+
+To achieve this goal the node must implement robust concurrent model. More specifically, all requests to the wallet
+web API endpoints must be handled concurrently.
+
+**TODO**: To think about the slowest endpoints, like `/api/v1/transactions`: is it possible to handle them
+fully concurrently? I mean publishing to the network, storing in the `wallet-db`, etc.
```
Comment (denisshevchenko) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: @adinapoli-iohk It's possible, right?

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+Initially clusters of nodes will be hosted by IOHK (on AWS EC2). And we need a mechanism of auto-scaling of the
+cluster: it must be able to grow proportionally to growing number of clients.
+
+**TODO**: To think about auto-scale supervisor. Suppose we have 1 cluster that consists of 100 nodes and each of node
+can reliably serve 100 clients. If supervisor sees that number of clients connected to this cluster is growing (and
+becomes more than 10000), it launches additional nodes.
+
+**TODO 2**: Launching additional node is relatively expensive action (in terms of time and memory). So it is better
+if one single node will be able to serve more clients than just increase a number of nodes in the cluster.
```
Comment (denisshevchenko) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: @adinapoli-iohk We have to benchmark it, to understand how many clients can be reliably served by one single node.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+When user wants to send money to other user, he creates a new payment. It produces a `POST`-request
+to `/api/v1/transactions/unsigned` endpoint, and response will contain a new transaction without a
+witness (signature).
+
+### Send Signed Transaction
+
+After user signs new transaction using its private keys (hosted on the client side) he sends it with
+a signature. It produces a `POST`-request to `/api/v1/transactions/signed`, node receives transaction
+with a signature, forms transaction witness and publish this transaction as usually.
+
```
Comment (denisshevchenko) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: @adinapoli-iohk Do we need some additional API endpoints for mobile clients?

File: docs/proposals/unified-backend-architecture/tech-spec.md (hunk 1 above)
Comment (domenkozar) [reply docs/proposals/unified-backend-architecture/tech-spec.md]: We might not base our autoscaling on number of users, but another resource. It's hard to say how it would scale before we run some benchmarks. It could be that bottleneck is CPU and you can't correlate that with number of subscribers.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+key), node creates a new wallet in `wallet-db` and sends response with information about new wallet.
+
+Even if node `N` which serves the client `C` will die, another node `M` (the client `C` will reconnect to) will be able
+to handle information about `C`'s wallet(s), because `wallet-db` is storing in the cluster.
+
+### Wallet Import/Export
+
+Wallet can be imported using client's secret key.
+
+**TODO**: To clarify this process. User can be able to obtain its secret key.
```
Comment (adinapoli-iohk) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Who can help us here, @vincenthz maybe? I guess this should become a little clearer once we finalise the Authentication scheme.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+Currently Daedalus can be treated as a heavyweight client: it works in a pair with Cardano node which stores
+full copy of the blockchain. It leads three problems:
+
+1. Such a heavyweight client must have enough disk space to store growing blockchain.
+2. Cardano node requires quite a lot system resources.
+3. During (first) start heavyweight client must sync (new) blocks and this is long process, even with good
+   internet connection.
+
+These problems make impossible to install such a heavyweight client on a mobile device. The purpose of Unified
+Backend Architecture is to solve these problems.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: While preserving a common architecture with the desktop client, and desktop client with hardware devices.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+We propose to create a lightweight Daedalus client. And it is possible to unify the different use cases under a
+single architecture, starting from the crucial observation that both interfacing with an hardware wallet or
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Mention the use cased earlier. Remember that linking to a YT ticket is not visible to 3rd party devs or auditors.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+The reader should be familiar with following concepts:
+
+1. Daedalus frontend,
+2. wallet backend web API,
+3. work with wallets,
+4. transaction processing.
+
+Also it is assumed that the reader knows basic terminology from [Cardano Docs](https://cardanodocs.com/),
+for example, UTXO.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Related work, to mention here or bibliography is the new wallet spec. Any new semi-formal spec for this use case will be based off of the new wallet spec.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+## Requirements
+
+Unification of the wallet backend architecture should allow to:
+
+1. work with different hardware wallets in the same manner,
+2. create lightweight Daedalus mobile clients (iOS/Android),
+3. create lightweight Daedalus desktop clients (Windows/macOS/Linux).
+
+The main benefit of such a lightweight client is independence from the growing blockchain: the client does not
+store blockchain locally and does not wait for a "Syncing blocks".
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Also ideally, the unified arch should also cover the existing desktop use case, running the wallet backend on the same machine. This is what will give us maximum sharing within the codebase.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+## Key Idea
+
+As mentioned before, the key idea is to unify the different use cases under a single architecture, we mean
+hardware wallets, mobile clients and lightweight desktop clients.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: And heavyweight desktop clients!  (full nodes running on the desktop)

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+
+In the case of [Ledger](https://www.ledgerwallet.com/) (or any other hardware wallet) the privte keys and the
+cryptography are hosted on the device, whereas in a mobile client they are held on a smartphone. And this schema
+fits for "featherweight Daedalus desktops" that do not want to host the growing blockchain on the computer's
+storage.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Yep.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+More specifically, a Daedalus mobile client will consist of an iOS or Android application which will communicate
+with a cluster of nodes that will keep the global UTXO state for the blockchain and that will be capable
+to reconstruct each mobile wallet state without hosting any user-specific data.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: :+1:

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+
+## Multi-Tenant Architecture
+
+Mobile clients will communicate with a clusters of nodes, so the basic requirement is an ability of one node to work
+with particular number of clients simultaneously. For example, if 1 cluster has to serve 10000 mobile clients and
+this cluster consists of 100 nodes, each node has to be able to work with 100 clients simultaneously and reliably.
+Thus, the longest procedure is a new payment (creating new transaction), and even if all these 100 clients will
+sign their transactions and send it for publishing at the same time, the node must be able to handle these
+transactions reliably and as fast as possible.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Yes, though actually it's the medium term average tx rate here that's important, short term fluctuations can be handled by queues/buffers. So there's the sustained tx submission rate and the queue size (to handle activity bursts).

File: docs/proposals/unified-backend-architecture/tech-spec.md (hunk 5 above)
Comment (dcoutts) [reply docs/proposals/unified-backend-architecture/tech-spec.md]: Yes, I think we should be able to achieve good enough throughput here, with the right design.

File: docs/proposals/unified-backend-architecture/tech-spec.md (hunk 1 above)
Comment (dcoutts) [reply docs/proposals/unified-backend-architecture/tech-spec.md]: Right, exact resource and tuning to do to be decided during detailed design analysis or later benchmarking.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+**TODO**: To think about auto-scale supervisor. Suppose we have 1 cluster that consists of 100 nodes and each of node
+can reliably serve 100 clients. If supervisor sees that number of clients connected to this cluster is growing (and
+becomes more than 10000), it launches additional nodes.
+
+_Domen said: We might not base our autoscaling on number of users, but another resource. It's hard to say how it
+would scale before we run some benchmarks. It could be that bottleneck is CPU and you can't correlate that with number
+of subscribers._
+
+**TODO 2**: Launching additional node is relatively expensive action (in terms of time and memory). So it is better
+if one single node will be able to serve more clients than just increase a number of nodes in the cluster.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Of course handling lots of clients with a single server is nice, and the biggest issue here is actually variability across operations/modes, e.g. it's no good if you can handle 10,000 established client sessions at once if you can only handle establishing 5 new client sessions at once. So as part of the design analysis, and deciding on what the important benchmarks are, we need to consider what the important scenarios are to benchmark so we can establish the real bottlnecks (e.g. for relays, it's the number of clients syncing concurrently that is the bottleneck, not the number just keeping up with the blockchain).

File: docs/proposals/unified-backend-architecture/tech-spec.md (hunk 8 above)
Comment (dcoutts) [reply docs/proposals/unified-backend-architecture/tech-spec.md]: Secret keys remain only on the client, so import is a frontend only operation (just turning the multi-word passphrase into the root keypair).

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+
+### Wallet Restoring
+
+Wallet can be restored. Use case: user changed a smartphone, install mobile Daedalus client and restore its wallet
+using backup phrase (for example, 12 words) generated during this wallet creation.
+
+Wallet restoring produces a `POST`-request to `/api/v1/wallets/?`, and node checks if such a wallet exists. If so,
+response contains an information about the wallet.
+
+**TODO**: To clarify this process.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Consult with delegation & restoration design options for details of recovery on backend without the backend having secret keys. Can either choose scheme with hardened or non-hardened keys. It's an interesting trade-off.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+
+## API Extension
+
+API v1 must be extended to support both the mobile clients and the hardware wallets.
+
+### Create Unsigned Transaction
+
+When user wants to send money to other user, he creates a new payment. It produces a `POST`-request
+to `/api/v1/transactions/unsigned` endpoint, and response will contain a new transaction without a
+witness (signature).
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Need to include in the response the info needed for the client or hardware to be able to sign the tx, so must include the HD wallet indexes of the keys needed to sign the tx. Unnecessary to contain tx hash since this can be derived (and would have to be checked anyway for security).

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+
+### Send Signed Transaction
+
+After user signs new transaction using its private keys (hosted on the client side) he sends it with
+a signature. It produces a `POST`-request to `/api/v1/transactions/signed`, node receives transaction
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Bikeshedding: probably different API endpoint name. Ask @adinapoli-iohk

File: docs/proposals/unified-backend-architecture/tech-spec.md (hunk 7 above)
Comment (dcoutts) [reply docs/proposals/unified-backend-architecture/tech-spec.md]: Not AFAIK, this should be the same flow for hardware or mobile clients.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+with a signature, forms transaction witness and publish this transaction as usually.
+
+## Wallet State
+
+An analysis of the computational complexity of rebuilding the state of a wallet (e.g. its balance and
+payments history) on the server in a way as much stateless as possible.
+
+**TODO**: To think about it.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Refer also to delegation workstream which is considering this point.

File: docs/proposals/unified-backend-architecture/tech-spec.md (hunk 3 above)
Comment (dcoutts) [reply docs/proposals/unified-backend-architecture/tech-spec.md]: Almost certainly we need to use deterministic (sequential or deterministically seeded PRNG).

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+A description of how the main operations on a crypto-wallet will be implemented, according to the final
+index derivation mechanism chosen (sequential vs random).
+
+**TODO**: To think about it.
+
+## Security and Possible Attacks
+
+### MITM
+
+To prevent MITM attacks we have to include a fee in transaction explicitly.
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Not prevent but mitigate the impact of. Also it's not just MITM, it's also simply a malicious/compromised wallet baclend (server).

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+
+Currently it is impossible, because transaction contains inputs, outputs and attributes, and we have to
+know _previous_ transaction to calculate a fee. So we must change transaction's format for explicit
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Correction: have to know current UTxO.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+including a fee in it.
+
+**TODO**: To think about it: should the fee be a part of transaction's attributes, or it should be a
+separate field?
```
Comment (dcoutts) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: To be defined in a separate spec focusing on that issue alone.

File: docs/proposals/unified-backend-architecture/tech-spec.md (hunk 2 above)
Comment (dcoutts) [reply docs/proposals/unified-backend-architecture/tech-spec.md]: This should eventually also be reviewed by the security internal auditors.

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+These are typical user stories (workflows), with description of API calls.
+
+### Initial State
+
+After installation of mobile client application user does not see any wallets, and he must create at least one to
+work with a client application.
+
+During the first connection with the node the new session is establishing, please read an explanation below.
+
+### Wallet Creation
```
Comment (adinapoli-iohk) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: I would rephrase the user stories in the usual style `As a user, I want to do X Y and Z to achieve W`. An example: https://www.romanpichler.com/blog/10-tips-writing-good-user-stories/ (It's quite comprehensive, but I guess you can extract the gist of it). This helps enormously because it helps you reason on the end goal of the user rather than the outcome (in this case the creation of a new wallet).

File: docs/proposals/unified-backend-architecture/tech-spec.md [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+
+User creates a new wallet. It produces a `POST`-request to `/api/v1/wallets/external` (with information about client's
+public key), node creates a new wallet in `wallet-db` and sends response with information about new wallet.
+
+Even if node `N` which serves the client `C` will die, another node `M` (the client `C` will reconnect to) will be able
+to handle information about `C`'s wallet(s), because `wallet-db` is storing in the cluster.
```
Comment (adinapoli-iohk) [inline docs/proposals/unified-backend-architecture/tech-spec.md]: Ok, let's reason about this one. We are a user and we want to create a wallet. In a normal scenario we would ask Daedalus to do this for us, feeding as input the 12 words mnemonic. Then calling `/api/v1/wallets` would give us back a new `Wallet` object, effectively creating it in the backend.

But in the context of the unified backend architecture (UBA) we do not want to share _any_ sensible information with the external world, so we won't send to `N` our 12 words mnemonic but just the _hash_ of the root public key for this HD wallet. In a nutshell, we would do client side what we are now doing on the server side. Let's note a couple of things:

* Incidentally, this was the original idea we had when we thought about the integration with external hardware wallets. If you take a look at [this endpoint](https://rebilly.github.io/ReDoc/?url=https://gist.githubusercontent.com/adinapoli-iohk/93791a963a9eb545315840c784fab3c2/raw/bbf320febe511839ce0545d669463fc13aee1742/stub.yaml#/paths/~1api~1wallets~1external/post), it pretty much accepts such _hash_ as input.

* If we were using random index derivation it wouldn't be safe to share such _hash_, as currently this is used (AFAIK) as a piece of secret for some symmetric encryption used to encrypt the random index into each `AddrAttributes` (I think that's called `HDPayload`).

Please @darko-mijic and @vincenthz tell me if I'm mistaken here. If I'm right, I would extend this user story to be a bit more precise and less vague 😁 

🚴 shedding: Let's come back to the choice of the name for the endpoint, as part of CHW-37.

**Comment by denisshevchenko** [comment]
@adinapoli-iohk I renamed user Stories in more human-friendly form. :-)

**Comment by jcmincke** [comment]
Closed because this PR is just too old.
